src/mjolnir/graphenhancer.cc

```diff
@@ -94,58 +96,232 @@ struct enhancer_stats {
+        !((directededge.forwardaccess() | directededge.reverseaccess()) & kVehicularAccess))
+      return false;
+
+    // try first the country state combo, then country only, then neither, then bail
+    auto found = tables.find(country_state_code);
+    if (found == tables.end())
+      found = tables.find(country_state_code.substr(0, 2));
+    if (found == tables.end())
+      found = tables.find("");
+    if (found == tables.end())
+      return false;
+
+    // urban or rural
+    const auto& speed_table = found->second[density <= 8];
+    size_t rc = static_cast<size_t>(directededge.classification());
+
+    // some kind of special use
+    switch (directededge.use()) {
+      case Use::kDriveway:
+        directededge.set_speed(speed_table.service[0]);
+        return true;
+      case Use::kAlley:
+        directededge.set_speed(speed_table.service[1]);
+        return true;
+      case Use::kParkingAisle:
+        directededge.set_speed(speed_table.service[2]);
+        return true;
+      case Use::kDriveThru:
+        directededge.set_speed(speed_table.service[3]);
+        return true;
+      default:
+        break;
+    }
+
+    // exit ramp
+    if (directededge.link()) {
+      // there are no unclassified links but there are residential
+      if (rc == static_cast<size_t>(RoadClass::kResidential))
+        rc = 5;
+      // these classes dont have links
+      if (rc >= speed_table.link_exiting.size())
+        return false;
+      // we use signage to tell if its an exit otherwise its just a link/ramp/turn channel
+      if (directededge.sign())
+        directededge.set_speed(speed_table.link_exiting[rc]);
+      else
+        directededge.set_speed(speed_table.link_turning[rc]);
+      return true;
+    }
+
+    // roundabout
+    if (directededge.roundabout()) {
+      directededge.set_speed(speed_table.roundabout[rc]);
+      return true;
+    }
+
+    // non-special use, just use the road class
+    directededge.set_speed(speed_table.way[rc]);
+    return true;
+  }
+
+  /**
+   * Update directed edge speed based on density and other edge parameters like
+   * surface type. TODO - add admin specific logic
+   * @param  directededge  Directed edge to update.
+   * @param  density       Relative road density.
+   * @param  urban_rc_speed Array of default speeds vs. road class for urban areas
+   * @param  rc_speed Array of default speeds vs. road class
+   * @param  infer_turn_channels flag indicating if we should infer tc.
+   * @param country_state_code 2 letter country and 2 letter state codes concatinated
+   *
+   */
+  void UpdateSpeed(DirectedEdge& directededge,
+                   const uint32_t density,
+                   const uint32_t* urban_rc_speed,
+                   bool infer_turn_channels,
+                   const std::string& country_state_code) {
+
+    // If we have config loaded we'll use that
+    if (!tables.empty() && FromConfig(directededge, density, country_state_code))
+      return;
```

Comment:
i highly recommend viewing this diff with whitespace hidden. if you dont do that it will look like i really modified a lot more than i did. what really happened here is i made this anonymous namespace function `UpdateSpeed` into a member function of `SpeedAssigner` other than indentation (and so also line wrapping) I didnt modify this function except to call the `FromConfig` method to first see if the config-driven speed assignment should be used.